### PR TITLE
Changes visualize driver functions to return graphviz object

### PR DESCRIPTION
**Pull request by skrawcz**
This change is backwards compatible. It returns and wires through returning the graphviz object created. That way in a notebook, it can render the DAG for you.

It also makes saving to a file optional -- you just need to pass in None explicitly. Adds test for it.

The hello world example is also updated with a real notebook and shows Thierry Jean's approach of using the autoreload extensions.

## Changes
 - returns graphviz object from display functions
 - changes hello world to include a proper notebook

## How I tested this
 - via notebook
 - unit tests

## Checklist

- [x] PR has an informative and human-readable title (this will be pulled into the release notes)
- [x] Changes are limited to a single goal (no scope creep)
- [x] Code passed the pre-commit check & code is left cleaner/nicer than when first encountered.
- [x] Any _change_ in functionality is tested
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future TODOs are captured in comments
- [x] Project documentation has been updated if adding/changing functionality.

### CODE REVIEW

**Comment by skrawcz** [comment]
Screen Shot 2023-01-24 at 9 54 16 PM

File: hamilton/driver.py [hunk 1]
```diff
@@ -359,7 +359,7 @@ def list_available_variables(self) -> List[Variable]:
     @capture_function_usage
     def display_all_functions(
         self, output_file_path: str, render_kwargs: dict = None, graphviz_kwargs: dict = None
-    ):
+    ) -> Optional["graphviz.Digraph"]:  # noqa F821
```
Comment (elijahbenizzy) [inline hamilton/driver.py:362]: Why optional?

File: hamilton/graph.py [hunk 2]
```diff
@@ -243,12 +243,12 @@ def display(
         output_file_path: str = "test-output/graph.gv",
         render_kwargs: dict = None,
         graphviz_kwargs: dict = None,
-    ):
+    ) -> Optional["graphviz.Digraph"]:  # noqa F821
```
Comment (elijahbenizzy) [inline hamilton/graph.py:246]: Not sure this one should be optional -- feel like its always nice to return and it won't break anything (at worst will make people's notebooks look better).

File: tests/test_graph.py [hunk 3]
```diff
@@ -530,6 +530,20 @@ def test_function_graph_display():
...
+    fg = graph.FunctionGraph(tests.resources.dummy_functions, config={"b": 1, "c": 2})
+    defined_nodes = set()
+    user_nodes = set()
+    for n in fg.get_nodes():
+        if n.user_defined:
+            user_nodes.add(n)
+        else:
+            defined_nodes.add(n)
+    digraph = fg.display(defined_nodes, user_nodes, None)
+    assert digraph is not None
```
Comment (elijahbenizzy) [inline tests/test_graph.py:544]: Should assert instance/something else here to be a little more specific IMO

**Review by elijahbenizzy — COMMENTED** [review]
LGTM -- a few minor points

File: hamilton/driver.py (hunk 1 above)
Comment (skrawcz) [reply hamilton/driver.py:362]: because if graphviz is not installed you will not get anything back.

File: hamilton/graph.py (hunk 2 above)
Comment (skrawcz) [reply hamilton/graph.py:246]: because if graphviz is not installed it wont return an object.

File: tests/test_graph.py (hunk 3 above)
Comment (skrawcz) [reply tests/test_graph.py:544]: checking the graph isn't valuable because there's already a test for that. But sure I can check the instance type.

**Review by elijahbenizzy — APPROVED** [review]

**Pull request merged**